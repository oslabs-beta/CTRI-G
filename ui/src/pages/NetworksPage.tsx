import { useState } from 'react';
import { createPortal } from 'react-dom';
import { useNavigate } from 'react-router-dom';
import Network from '../components/Network';
import NetworkForm from '../components/NetworkForm';
import FormModal from '../components/container-form/FormModal';
import { ContainerInfo, NetworkInfo } from '../interfaces/interfaces';

const NetworksPage = (props: {
  networks: NetworkInfo[] | [];
  containers: ContainerInfo[] | [];
}) => {
  const nav = useNavigate();
  const networkEl: JSX.Element[] = [];
  const hostNone: JSX.Element[] = [];
  const defaultBridge: JSX.Element[] = [];

  //function to display 'Add Network' popup
  // const showAddNetworkForm = () => {
  //   const addNetworkForm = document.getElementById('addNetworkForm');
  //   if (addNetworkForm !== null) {
  //     addNetworkForm.style.display = 'flex';
  //   }
  // };

  const [displayAddNetworkForm, setDisplayAddNetworkForm] = useState(false);

  const closeAddNetworkForm = () => {
    setDisplayAddNetworkForm(false);
  };

  //iterates through networks and creates a container for each
  props.networks.forEach((network, i: number) => {
    const newEl = (
      <Network
        key={`network${i}`}
        networkIndex={i}
        network={network}
        containers={props.containers}
        id={
          network.Driver !== 'bridge' || network.Name === 'bridge'
            ? 'defaultNetwork'
            : 'userNetwork'
        }
        allNetworks={props.networks}
      />
    );
    //pushes new element into specified array
    if (network.Name === 'host' || network.Name === 'none') {
      hostNone.push(newEl);
    } else if (network.Name === 'bridge') {
      defaultBridge.push(newEl);
    } else {
      networkEl.push(newEl);
    }
  });

  //combine arrays in desired order
  networkEl.push(...hostNone);
  networkEl.unshift(...defaultBridge);

  return (
    <div className='mainContainer'>
      <div className='buttonContainer'>
        <button className='button' onClick={() => nav('Visualizer')}>
          Visualizer
        </button>
      </div>
      <div className='hostContainer'>
        <img
          className='comp'
          src={
            'https://res.cloudinary.com/dbinuhocd/image/upload/v1692796941/computer-icon_zns1lr.png'
          }
          alt='computer-icon'
        />
      </div>
      <div className='addNetworkButtonContainer'>
        {createPortal(
          <FormModal open={displayAddNetworkForm} onClose={closeAddNetworkForm}>
            <NetworkForm closeAddNetworkForm={closeAddNetworkForm} />
          </FormModal>,
          document.body,
        )}
        <button
          className='button addNetworkButton'
          onClick={() => setDisplayAddNetworkForm(true)}>
          Add Network
        </button>
      </div>
      <div className='networksContainer'>{networkEl}</div>
    </div>
  );
};

export default NetworksPage;
