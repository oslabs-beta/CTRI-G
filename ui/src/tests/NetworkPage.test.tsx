//add imports here
import NetworksPage from '../pages/NetworksPage';
import { render, screen, fireEvent } from '@testing-library/react';
import { create } from 'react-test-renderer';
import '@testing-library/jest-dom';

// mocks go here if needed
jest.mock('@docker/extension-api-client', () => ({
  createDockerDesktopClient: jest.fn().mockReturnValue({
    docker: {
      cli: {
        exec: jest.fn(),
      },
    },
    desktopUI: {
      toast: {
        error: jest.fn(),
        success: jest.fn(),
        warning: jest.fn(),
      },
    },
  }),
}));

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

// set up NetworkPage tests, clearing mocks and restarting modules before each
describe('NetworkPage unit tests', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetModules();
  });
  describe('Network', () => {
    beforeEach(() => {
      jest.clearAllMocks();
      jest.resetModules();
    });
    // description of props goes here
    const userCreatedNetworks = {
      networks: [
        {
          Driver: 'bridge',
          Name: 'Bridge',
          ID: 'A',
        },
        {
          Driver: 'host',
          Name: 'host',
          ID: 'B',
        },
        {
          Driver: 'none',
          Name: 'null',
          ID: 'C',
        },
        {
          Driver: 'bridge',
          Name: 'Eg1',
          ID: 'D',
        },
        {
          Driver: 'bridge',
          Name: 'Eg2',
          ID: 'E',
        },
        {
          Driver: 'bridge',
          Name: 'Eg3',
          ID: 'F',
        },
      ],
    };

    const containers = [
      {
        Name: 'containertest',
        Id: 'rjsfjs',
        Image: 'containertest',
        State: 'running',
        Networks: 'none',
        Ports: [],
      },
    ];
    const noUserCreatedNetworksProps = {
      networks: [
        {
          Driver: 'bridge',
          Name: 'Bridge',
          ID: 'A',
        },
        {
          Driver: 'host',
          Name: 'host',
          ID: 'B',
        },
        {
          Driver: 'none',
          Name: 'null',
          ID: 'C',
          Containers: ['Port Navigator'],
        },
      ],
      containers: [
        {
          Id: 'abc',
          Image: 'portNavigator',
          Name: 'Port Navigator',
          Networks: 'Test Network',
          Ports: [],
          State: 'Test',
        },
      ],
    };
    // Should display a containers (later a visualizer) button
    test('Should display a Visualizer button on initial render', () => {
      render(<NetworksPage {...noUserCreatedNetworksProps} />);
      const visualizerButton = screen.getByRole('button', {
        name: /Visualizer/,
      });
      expect(visualizerButton).toBeDefined();
    });

    // Should switch to containers/visualizer page when containers/visualizer button is clicked
    // test('Should switch to Visualizer page when Visualizer button is clicked', () => {
    //   render(<NetworksPage {...noUserCreatedNetworksProps} />);
    //   const visualizerButton = screen.getByRole('button', {
    //     name: /Visualizer/,
    //   });
    //   fireEvent.click(visualizerButton);
    //   const addNetworkButton = screen.getByRole('button', {
    //     name: /Add Network/,
    //   });
    //   expect(addNetworkButton).not.toBeDefined();
    // });

    // Containers/visualizer button should switch to Networks button when Containers/visualizer button is clicked
    // test('Should switch Visualizer button to Networks button when Visualizer button is clicked', () => {
    //   render(<NetworksPage {...noUserCreatedNetworksProps} />);
    //   const visualizerButton = screen.getByRole('button', {
    //     name: /Visualizer/,
    //   });
    //   fireEvent.click(visualizerButton);
    //   const networksButton = screen.getByRole('button', {
    //     name: /Networks/,
    //   });
    //   expect(networksButton).toBeDefined();
    // });

    // Should switch to networks page when Networks button is clicked
    // test('Should switch to Networks page when Networks button is clicked', () => {
    //   render(<NetworksPage {...noUserCreatedNetworksProps} />);
    //   const visualizerButton = screen.getByRole('button', {
    //     name: /Visualizer/,
    //   });
    //   fireEvent.click(visualizerButton);
    //   const addNetworkButton = screen.getByRole('button', {
    //     name: /Add Network/,
    //   });
    //   expect(addNetworkButton).toBeDefined();
    // });

    // Should display an Add Network button
    test('Should display an Add Network button', () => {
      render(<NetworksPage {...noUserCreatedNetworksProps} />);
      const addNetworkButton = screen.getByRole('button', {
        name: /Add Network/,
      });
      expect(addNetworkButton).toBeDefined();
    });

    // Should display Add Network Form on click of Add Network button
    test('Should display an Add Network form when the Add Network button is clicked', () => {
      render(<NetworksPage {...noUserCreatedNetworksProps} />);
      const addNetworkButton = screen.getByRole('button', {
        name: /Add Network/,
      });
      fireEvent.click(addNetworkButton);
      const addNetworkFormText = screen.getByText(/Advanced network settings/);
      expect(addNetworkFormText).toBeVisible();
    });

    // Should display a close Add Network Form button
    test('Should display an Add Network Form Close button when the Add Network button is clicked', () => {
      render(<NetworksPage {...noUserCreatedNetworksProps} />);
      const addNetworkButton = screen.getByRole('button', {
        name: /Add Network/,
      });
      fireEvent.click(addNetworkButton);
      const addNetworkFormCloseButtonArray = screen.queryAllByRole('button', {
        name: /x/,
      });
      expect(addNetworkFormCloseButtonArray.length).toEqual(4);
    });

    // Should close Add Network Form on click of Close Add Network Form button
    test('Should closelose Add Network form when the Close Add Network Form button is clicked', () => {
      render(<NetworksPage {...noUserCreatedNetworksProps} />);
      const addNetworkButton = screen.getByRole('button', {
        name: /Add Network/,
      });
      fireEvent.click(addNetworkButton);
      const addNetworkFormCloseButtonArray = screen.queryAllByText(/x/);
      fireEvent.click(addNetworkFormCloseButtonArray[3]);
      const addNetworkFormText = screen.queryByText(
        'Advanced network settings'
      );
      expect(addNetworkFormText).toBeNull();
    });

    // Should display an Image at the top (not a network)
    // test('Should display an Image at the top of the page', () => {
    //   render(<NetworksPage {...noUserCreatedNetworksProps} />);
    //   const Host = screen.getByText(/Host/);
    //   expect(Host).toBeDefined();
    // });

    // Should display a bridge network, a host network, and a none network if no user created networks are passed in as props
    test('renders bridge network if no user created props are passed in as props', () => {
      render(<NetworksPage {...noUserCreatedNetworksProps} />);
      const networkName = screen.getByText(/Bridge/);
      expect(networkName).toBeInTheDocument();
    });

    test('renders none network if no user created props are passed in as props', () => {
      render(<NetworksPage {...noUserCreatedNetworksProps} />);
      const noneNetworkText = screen.getByText(/none/);
      expect(noneNetworkText).toBeDefined();
    });

    test('renders host network if no user created props are passed in as props', () => {
      render(<NetworksPage {...noUserCreatedNetworksProps} />);
      const hostTextArray = screen.queryAllByText(/host/);
      expect(hostTextArray.length).toEqual(2);
    });

    // Should display multiple user-created networks in addition to bridge, host, and none networks if multiple user-created networks are passed in as props
    test('renders none network if it is provided  as props', () => {
      render(<NetworksPage {...userCreatedNetworks} containers={containers} />);
      const noneNetworkText = screen.getByText(/none/);
      expect(noneNetworkText).toBeDefined();
    });

    test('renders host network if it is provided  as props', () => {
      render(<NetworksPage {...userCreatedNetworks} containers={containers} />);
      const hostTextArray = screen.queryAllByText(/host/);
      expect(hostTextArray.length).toEqual(2);
    });

    test('renders bridge network if it is provided  as props', () => {
      render(<NetworksPage {...userCreatedNetworks} containers={containers} />);
      const networkName = screen.getByText(/Bridge/);
      expect(networkName).toBeInTheDocument();
    });

    test('renders user created network if it is provided  as props', () => {
      render(<NetworksPage {...userCreatedNetworks} containers={containers} />);
      const networkName = screen.getByText(/Eg1/);
      expect(networkName).toBeInTheDocument();
    });

    test('renders a second user created network if it is provided  as props', () => {
      render(<NetworksPage {...userCreatedNetworks} containers={containers} />);
      const networkName = screen.getByText(/Eg2/);
      expect(networkName).toBeInTheDocument();
    });

    test('renders a third user created network if it is provided  as props', () => {
      render(<NetworksPage {...userCreatedNetworks} containers={containers} />);
      const networkName = screen.getByText(/Eg3/);
      expect(networkName).toBeInTheDocument();
    });

    test('renders NetworkPage', () => {
      const networksPage = create(
        <NetworksPage {...userCreatedNetworks} containers={containers} />
      ).toJSON();
      expect(networksPage).toMatchSnapshot();
    });
  });
});
