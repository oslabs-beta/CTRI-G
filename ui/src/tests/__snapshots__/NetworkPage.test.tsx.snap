// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkPage unit tests Network renders NetworkPage 1`] = `
<div
  className="mainContainer"
>
  <div
    className="buttonContainer"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Visualizer
    </button>
  </div>
  <div
    className="hostContainer"
  >
    <img
      alt="computer-icon"
      className="comp"
      src="https://res.cloudinary.com/dbinuhocd/image/upload/v1692796941/computer-icon_zns1lr.png"
    />
  </div>
  <div
    className="addNetworkButtonContainer"
  >
    <button
      className="addNetworkButton"
      onClick={[Function]}
    >
      Add Network
    </button>
  </div>
  <div
    className="networksContainer"
  >
    <div
      className="userNetwork"
    >
      <div
        className="networkContainer"
      >
        <div
          className="networkLabel"
        >
          <strong>
            Network: 
          </strong>
          Bridge
        </div>
        <hr
          className="hrNet"
        />
        <div
          className="networkInfo"
        >
          <div
            className="Driver"
          >
            <strong>
              Driver: 
            </strong>
            bridge
          </div>
          <div
            className="Gateway"
          >
            <strong>
              Gateway: 
            </strong>
            null
          </div>
          <div
            className="Subnet"
          >
            <strong>
              Subnet: 
            </strong>
            null
          </div>
        </div>
        <hr />
        <div
          className="containerNetworkFeatures"
        >
          <div
            className="connectedContainerContainer"
          >
            <p
              className="connectedText"
            >
              Connected
              <br />
              Containers
            </p>
            <p
              className="divider"
            >
              |
            </p>
            <p
              className="connectedCount"
            />
          </div>
          <div />
          <button
            className="innerButton"
            id="0ShowHideNetworksButton"
            onClick={[Function]}
          >
            Add Container
          </button>
        </div>
        <button
          className="deleteNetworkButton"
          onClick={[Function]}
        >
          x
        </button>
      </div>
      <div
        className="containersContainer"
        id="0ContainersContainer"
        style={
          {
            "display": "none",
          }
        }
      />
    </div>
    <div
      className="defaultNetwork"
    >
      <div
        className="networkContainer"
      >
        <div
          className="networkLabel"
        >
          <strong>
            Network: 
          </strong>
          null
        </div>
        <hr
          className="hrNet"
        />
        <div
          className="networkInfo"
        >
          <div
            className="Driver"
          >
            <strong>
              Driver: 
            </strong>
            none
          </div>
          <div
            className="Gateway"
          >
            <strong>
              Gateway: 
            </strong>
            null
          </div>
          <div
            className="Subnet"
          >
            <strong>
              Subnet: 
            </strong>
            null
          </div>
        </div>
        <hr />
        <div
          className="containerNetworkFeatures"
        >
          <div
            className="connectedContainerContainer"
          >
            <p
              className="connectedText"
            >
              Connected
              <br />
              Containers
            </p>
            <p
              className="divider"
            >
              |
            </p>
            <p
              className="connectedCount"
            />
          </div>
          <div />
          <button
            className="innerButton"
            id="2ShowHideNetworksButton"
            onClick={[Function]}
          >
            Add Container
          </button>
        </div>
        <button
          className="deleteNetworkButton"
          onClick={[Function]}
        >
          x
        </button>
      </div>
      <div
        className="containersContainer"
        id="2ContainersContainer"
        style={
          {
            "display": "none",
          }
        }
      />
    </div>
    <div
      className="userNetwork"
    >
      <div
        className="networkContainer"
      >
        <div
          className="networkLabel"
        >
          <strong>
            Network: 
          </strong>
          Eg1
        </div>
        <hr
          className="hrNet"
        />
        <div
          className="networkInfo"
        >
          <div
            className="Driver"
          >
            <strong>
              Driver: 
            </strong>
            bridge
          </div>
          <div
            className="Gateway"
          >
            <strong>
              Gateway: 
            </strong>
            null
          </div>
          <div
            className="Subnet"
          >
            <strong>
              Subnet: 
            </strong>
            null
          </div>
        </div>
        <hr />
        <div
          className="containerNetworkFeatures"
        >
          <div
            className="connectedContainerContainer"
          >
            <p
              className="connectedText"
            >
              Connected
              <br />
              Containers
            </p>
            <p
              className="divider"
            >
              |
            </p>
            <p
              className="connectedCount"
            />
          </div>
          <div />
          <button
            className="innerButton"
            id="3ShowHideNetworksButton"
            onClick={[Function]}
          >
            Add Container
          </button>
        </div>
        <button
          className="deleteNetworkButton"
          onClick={[Function]}
        >
          x
        </button>
      </div>
      <div
        className="containersContainer"
        id="3ContainersContainer"
        style={
          {
            "display": "none",
          }
        }
      />
    </div>
    <div
      className="userNetwork"
    >
      <div
        className="networkContainer"
      >
        <div
          className="networkLabel"
        >
          <strong>
            Network: 
          </strong>
          Eg2
        </div>
        <hr
          className="hrNet"
        />
        <div
          className="networkInfo"
        >
          <div
            className="Driver"
          >
            <strong>
              Driver: 
            </strong>
            bridge
          </div>
          <div
            className="Gateway"
          >
            <strong>
              Gateway: 
            </strong>
            null
          </div>
          <div
            className="Subnet"
          >
            <strong>
              Subnet: 
            </strong>
            null
          </div>
        </div>
        <hr />
        <div
          className="containerNetworkFeatures"
        >
          <div
            className="connectedContainerContainer"
          >
            <p
              className="connectedText"
            >
              Connected
              <br />
              Containers
            </p>
            <p
              className="divider"
            >
              |
            </p>
            <p
              className="connectedCount"
            />
          </div>
          <div />
          <button
            className="innerButton"
            id="4ShowHideNetworksButton"
            onClick={[Function]}
          >
            Add Container
          </button>
        </div>
        <button
          className="deleteNetworkButton"
          onClick={[Function]}
        >
          x
        </button>
      </div>
      <div
        className="containersContainer"
        id="4ContainersContainer"
        style={
          {
            "display": "none",
          }
        }
      />
    </div>
    <div
      className="userNetwork"
    >
      <div
        className="networkContainer"
      >
        <div
          className="networkLabel"
        >
          <strong>
            Network: 
          </strong>
          Eg3
        </div>
        <hr
          className="hrNet"
        />
        <div
          className="networkInfo"
        >
          <div
            className="Driver"
          >
            <strong>
              Driver: 
            </strong>
            bridge
          </div>
          <div
            className="Gateway"
          >
            <strong>
              Gateway: 
            </strong>
            null
          </div>
          <div
            className="Subnet"
          >
            <strong>
              Subnet: 
            </strong>
            null
          </div>
        </div>
        <hr />
        <div
          className="containerNetworkFeatures"
        >
          <div
            className="connectedContainerContainer"
          >
            <p
              className="connectedText"
            >
              Connected
              <br />
              Containers
            </p>
            <p
              className="divider"
            >
              |
            </p>
            <p
              className="connectedCount"
            />
          </div>
          <div />
          <button
            className="innerButton"
            id="5ShowHideNetworksButton"
            onClick={[Function]}
          >
            Add Container
          </button>
        </div>
        <button
          className="deleteNetworkButton"
          onClick={[Function]}
        >
          x
        </button>
      </div>
      <div
        className="containersContainer"
        id="5ContainersContainer"
        style={
          {
            "display": "none",
          }
        }
      />
    </div>
    <div
      className="defaultNetwork"
    >
      <div
        className="networkContainer"
      >
        <div
          className="networkLabel"
        >
          <strong>
            Network: 
          </strong>
          host
        </div>
        <hr
          className="hrNet"
        />
        <div
          className="networkInfo"
        >
          <div
            className="Driver"
          >
            <strong>
              Driver: 
            </strong>
            host
          </div>
          <div
            className="Gateway"
          >
            <strong>
              Gateway: 
            </strong>
            null
          </div>
          <div
            className="Subnet"
          >
            <strong>
              Subnet: 
            </strong>
            null
          </div>
        </div>
        <hr />
        <div
          className="containerNetworkFeatures"
        >
          <div
            className="connectedContainerContainer"
          >
            <p
              className="connectedText"
            >
              Connected
              <br />
              Containers
            </p>
            <p
              className="divider"
            >
              |
            </p>
            <p
              className="connectedCount"
            />
          </div>
          <div />
          <button
            className="innerButton"
            id="1ShowHideNetworksButton"
            onClick={[Function]}
          >
            Add Container
          </button>
        </div>
        <button
          className="deleteNetworkButton"
          onClick={[Function]}
        >
          x
        </button>
      </div>
      <div
        className="containersContainer"
        id="1ContainersContainer"
        style={
          {
            "display": "none",
          }
        }
      />
    </div>
  </div>
</div>
`;
